### PR TITLE
Add Windows support for the Zabbix Agent

### PR DESCRIPTION
**Pull request by wardhus**
Hi,

i've added support for the Zabbix agent on Windows.

### CODE REVIEW

File: manifests/agent.pp [hunk 1]
```diff
@@ -272,7 +273,7 @@
   $tlsservercertsubject                   = $zabbix::params::agent_tlsservercertsubject,
   String $agent_config_owner              = $zabbix::params::agent_config_owner,
   String $agent_config_group              = $zabbix::params::agent_config_group,
-  Boolean $manage_selinux                 = $zabbix::params::manage_selinux,
+  Optional[Boolean] $manage_selinux       = $zabbix::params::manage_selinux,
```
Comment (bastelfreak) [inline manifests/agent.pp]: Why did you make this optional? Wouldn't it make more sense to default to false on Windows?

File: manifests/agent.pp [hunk 2]
```diff
@@ -368,7 +370,7 @@
     }
   }
 
-  if $agent_configfile_path != '/etc/zabbix/zabbix_agentd.conf' {
+  if $agent_configfile_path != '/etc/zabbix/zabbix_agentd.conf' and $::kernel != 'windows' {
```
Comment (bastelfreak) [inline manifests/agent.pp]: the topscope facts are deprecated. Please use:
```suggestion
  if $agent_configfile_path != '/etc/zabbix/zabbix_agentd.conf' and $facts['kernel'] != 'windows' {
```

File: manifests/agent.pp [hunk 3]
```diff
@@ -395,7 +397,7 @@
     ensure  => present,
     owner   => $agent_config_owner,
     group   => $agent_config_group,
-    mode    => '0644',
+    mode    => '0664',
```
Comment (bastelfreak) [inline manifests/agent.pp]: can you explain why the change is needed?

File: metadata.json [hunk 4]
```diff
@@ -48,6 +48,10 @@
       "name": "puppet/selinux",
       "version_requirement": ">= 1.1.0 < 2.0.0"
     }
+    {
+      "name": "chocolatey-chocolatey",
```
Comment (bastelfreak) [inline metadata.json]: In my opinion this is a optional dependency and should not be added to the metadata.json. Can you please document this in the README.md?

File: metadata.json [hunk 5]
```diff
@@ -146,6 +150,9 @@
     },
     {
       "operatingsystem": "Gentoo"
+    },
+    {
+      "operatingsystem": "Windows"
```
Comment (bastelfreak) [inline metadata.json]: Can you add the specific windows versions it works on?

**Comment by bastelfreak** [comment]
Hi @wardhus, thanks for the work! Can you please take a look at the inline comments, rebase gainst our latest master and add some rspec tests?

File: manifests/agent.pp (hunk 3 above)
Comment (wardhus) [reply manifests/agent.pp]: This is needed because Puppet and Windows file permissions don't play nice. If I leave it at 644 Puppet will change it every run (but nothing will actually happen)

File: manifests/agent.pp (hunk 1 above)
Comment (wardhus) [reply manifests/agent.pp]: I made this undef because the fact $facts['selinux'] is undef on windows. I could move manage_selinux to every OS in the case block, but this just seems more confusing to me.

File: metadata.json [hunk 6]
```diff
@@ -146,6 +146,15 @@
     },
     {
       "operatingsystem": "Gentoo"
+    },
+    {
+      "operatingsystem": "windows",
+      "operatingsystemrelease": [
+        "2008R2",
```
Comment (bastelfreak) [inline metadata.json]: Not a requirement to get this merged, but: Could you have a look at https://github.com/camptocamp/facterdb/tree/master/facts and maybe contribute some facts from windows machines? This repo basically has copies from the `facter -j` output. This is used in our tests to mock facts. It would really help us to get some more windows facts.

File: manifests/agent.pp (hunk 1 above)
Comment (bastelfreak) [reply manifests/agent.pp]: oh crap. It seems to always be true/false on Linux. I propose the following change for the params.pp:

```puppet
$manage_selinux = $facts['selinux'] ? {
  true     => true,
  default => false,
}
```

**Comment by bastelfreak** [comment]
Hey. This still needs a rebase against our latest master. You can join our slack channel at https://slack.puppet.com or our IRC channel #voxpupuli on freenode if you need any help.

**Comment by JasonMcMahan** [comment]
Is this on the road map to add the ability to install zabbix agents on windows?

**Comment by bastelfreak** [comment]
@JasonMcMahan we cannot work it in the current state. If you're interested in it, you can clone the branch, rebase it and address the issues from the comments, and submit it as a new pull request.

**Comment by bastelfreak** [comment]
Ping @wardhus :)

**Comment by bastelfreak** [comment]
I'm going to close this PR since it didn't receive a lot of attention in the past months. If anybody is interested in this please feel free to reopen and rebase it.
